tests/integration/test_credentials.py

```diff
@@ -0,0 +1,106 @@
+# Copyright 2015 Amazon.com, Inc. or its affiliates. All Rights Reserved.
+#
+# Licensed under the Apache License, Version 2.0 (the "License"). You
+# may not use this file except in compliance with the License. A copy of
+# the License is located at
+#
+# http://aws.amazon.com/apache2.0/
+#
+# or in the "license" file accompanying this file. This file is
+# distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF
+# ANY KIND, either express or implied. See the License for the specific
+# language governing permissions and limitations under the License.
+
+import os
+import mock
+
+from botocore.session import Session
+from tests import unittest
+
+
+class TestCredentialPrecedence(unittest.TestCase):
+    def setUp(self):
+        # Clean up any existing environment
+        for name in ['AWS_ACCESS_KEY_ID',
+                     'AWS_SECRET_ACCESS_KEY',
+                     'BOTO_DEFAULT_PROFILE']:
+            if name in os.environ:
+                del os.environ[name]
+
+        # Set the config file to something that doesn't exist, then
+        # set the shared credential file to our test config.
+        os.environ['AWS_CONFIG_FILE'] = '~/.aws/config-missing'
+        Session.SessionVariables['credentials_file'] = (
+            None, None,
+            os.path.join(os.path.dirname(__file__), 'test-credentials'))
+
+    def test_access_secret_vs_profile_env(self):
+        # If all three are given, then the access/secret keys should
+        # take precedence.
+        os.environ['AWS_ACCESS_KEY_ID'] = 'env'
+        os.environ['AWS_SECRET_ACCESS_KEY'] = 'env-secret'
+        os.environ['BOTO_DEFAULT_PROFILE'] = 'test'
+
+        s = Session()
+        credentials = s.get_credentials()
+
+        self.assertEqual(credentials.access_key, 'env')
+        self.assertEqual(credentials.secret_key, 'env-secret')
+
+    @mock.patch('botocore.credentials.Credentials')
```

Comment:
I'm not really a fan of unnecessary patching, *especially* when it's not necessary.  Why can't we just check the credential object directly?